src/main/java/ch/njol/skript/util/PersistentDataUtils.java

```diff
@@ -0,0 +1,340 @@
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  Skript is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ *
+ * Copyright 2011-2017 Peter Güttinger and contributors
+ */
+package ch.njol.skript.util;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.regex.Pattern;
+
+import org.bukkit.NamespacedKey;
+import org.bukkit.block.Block;
+import org.bukkit.block.TileState;
+import org.bukkit.inventory.meta.BlockStateMeta;
+import org.bukkit.inventory.meta.ItemMeta;
+import org.bukkit.metadata.FixedMetadataValue;
+import org.bukkit.metadata.MetadataValue;
+import org.bukkit.metadata.Metadatable;
+import org.bukkit.persistence.PersistentDataAdapterContext;
+import org.bukkit.persistence.PersistentDataHolder;
+import org.bukkit.persistence.PersistentDataType;
+
+import org.eclipse.jdt.annotation.Nullable;
+
+import ch.njol.skript.Skript;
+import ch.njol.skript.aliases.ItemType;
+import ch.njol.skript.registrations.Classes;
+import ch.njol.skript.variables.SerializedVariable.Value;
+
+/**
+ * @author APickledWalrus
+ */
+public class PersistentDataUtils {
+
+	private final static PersistentDataType<byte[], Value> SKRIPT_TYPE = new SkriptDataType();
+
+	/*
+	 * General Utility Methods
+	 */
+
+	/**
+	 * For a {@linkplain Block} or an {@linkplain ItemType}, only parts of them are actually a {@linkplain PersistentDataHolder}.
+	 * This gets the 'actual' holder from those types (e.g. ItemMeta or TileState).
+	 * @param holder A {@linkplain PersistentDataHolder}, a {@linkplain Block}, or an {@linkplain ItemType}.
+	 * @return The actual {@linkplain PersistentDataHolder}, or null if the object's actual holder can't be found.
+	 */
+	@Nullable
+	private static PersistentDataHolder getActualHolder(@Nullable Object holder) {
+		if (holder == null) {
+			return null;
+		} else if (holder instanceof PersistentDataHolder) {
+			return (PersistentDataHolder) holder;
+		} else if (holder instanceof ItemType) {
+			return ((ItemType) holder).getItemMeta();
+		} else if (holder instanceof Block) {
+			if (((Block) holder).getState() instanceof TileState)
+				return ((TileState) ((Block) holder).getState());
+		}
+		return null;
+	}
+
+	/**
+	 * From:
+	 * https://hub.spigotmc.org/stash/projects/SPIGOT/repos/bukkit/browse/src/main/java/org/bukkit/NamespacedKey.java#33
+	 */
+	@SuppressWarnings("null")
+	private static final Pattern VALID_KEY = Pattern.compile("[a-z0-9/._-]+");
+
+	private static boolean isValidKey(String key) {
```

Comment:
instead of copying the pattern, why not just catch the exception thrown by the NamespacedKey constructor

Reply:
What do you think about it following this order:

get/set/etc methods:
  method to get the namespaced key
    do replacement stuff
    try to construct namespaced key
      return the key if it's valid - if it isn't valid, catch the exception, print console warning, return null